fast64_internal/f3d/f3d_material.py

```diff
@@ -1271,467 +1334,380 @@ def update_node_values_of_material(material, context):
     else:
-        nodes["Cycle Type"].outputs[0].default_value = (
-            1 if material.rdp_settings.g_mdsft_cycletype == "G_CYC_2CYCLE" else 0
-        )
-        nodes["Cull Front"].outputs[0].default_value = 0 if material.rdp_settings.g_cull_front else 1
-        nodes["Cull Back"].outputs[0].default_value = 0 if material.rdp_settings.g_cull_back else 1
-        nodes["Texture Gen"].outputs[0].default_value = 1 if material.rdp_settings.g_tex_gen else 0
-        nodes["Texture Gen Linear"].outputs[0].default_value = 1 if material.rdp_settings.g_tex_gen_linear else 0
-        nodes["Shading"].outputs[0].default_value = 0 if not material.rdp_settings.g_shade else 1
-        nodes["Lighting"].outputs[0].default_value = 0 if not material.rdp_settings.g_lighting else 1
-
-        if material.use_default_lighting:
-            nodes["Ambient Color"].outputs[0].default_value = (
-                material.default_light_color[0],
-                material.default_light_color[1],
-                material.default_light_color[2],
-                material.default_light_color[3],
-            )
+        col = [1.0, 1.0, 1.0, 1.0]
+        amb_col = [0.5, 0.5, 0.5, 1.0]
+        nodes["Shade Color"].inputs["Shade Color"].default_value = tuple(c for c in col)
+        nodes["Shade Color"].inputs["Ambient Color"].default_value = tuple(c for c in amb_col)
+        link_if_none_exist(material, nodes["ShadeColOut"].outputs[0], nodes["Shade Color"].inputs["Shade Color"])
+        link_if_none_exist(material, nodes["AmbientColOut"].outputs[0], nodes["Shade Color"].inputs["Ambient Color"])
+
+
+def update_color_node(combiner_inputs, color: Color, prefix: str):
+    """Function for updating either Prim or Env colors"""
+    # TODO: feature to toggle gamma correction
+    corrected_prim = gammaCorrect(color)
+    combiner_inputs[f"{prefix} Color"].default_value = (
+        corrected_prim[0],
+        corrected_prim[1],
+        corrected_prim[2],
+        1.0,
+    )
+    combiner_inputs[f"{prefix} Alpha"].default_value = color[3]
+
+
+# prim_color | Prim
+# env_color | Env
+def get_color_input_update_callback(attr_name="", prefix=""):
+    def input_update_callback(self: bpy.types.Material, context: bpy.types.Context):
+        with F3DMaterial_UpdateLock(get_material_from_context(context)) as material:
+            if not material:
+                return
+            f3dMat: "F3DMaterialProperty" = material.f3d_mat
+            nodes = material.node_tree.nodes
+            combiner_inputs = nodes["CombinerInputs"].inputs
+            update_color_node(combiner_inputs, getattr(f3dMat, attr_name), prefix)
+
+    return input_update_callback
+
+
+def update_node_values_of_material(material: bpy.types.Material, context):
+    nodes = material.node_tree.nodes
+
+    # Case where f3d render engine is used instead of node graph
+    # Note that v4 doesn't change the node graph from v3, so we use that name
```

Comment:
how relevant is this still?

Reply:
comment is outdated, but its good to check that the material is actually an f3d structure

Reply:
removed the comment and changed `hasNodeGraph` to `has_f3d_nodes`